frontend/appflowy_flutter/lib/plugins/database/application/cell/cell_controller.dart

```diff
@@ -35,48 +50,52 @@ class CellController<T, D> extends Equatable {
   SingleFieldListener? _fieldListener;
   CellDataNotifier<T?>? _cellDataNotifier;
 
-  VoidCallback? _onCellFieldChanged;
+  void Function(FieldPB field)? _onCellFieldChanged;
   VoidCallback? _onRowMetaChanged;
   Timer? _loadDataOperation;
   Timer? _saveDataOperation;
 
-  String get viewId => _cellContext.viewId;
-
   RowId get rowId => _cellContext.rowId;
-
-  String get fieldId => _cellContext.fieldInfo.id;
-
-  FieldInfo get fieldInfo => _cellContext.fieldInfo;
-
-  FieldType get fieldType => _cellContext.fieldInfo.fieldType;
-
-  String? get emoji => _cellContext.emoji;
+  String get fieldId => _cellContext.fieldId;
+  FieldInfo get fieldInfo => _fieldController.getField(_cellContext.fieldId)!;
+  FieldType get fieldType =>
+      _fieldController.getField(_cellContext.fieldId)!.fieldType;
+  RowMetaPB? get rowMeta => _rowCache.getRow(rowId)?.rowMeta;
+  String? get icon => rowMeta?.icon;
+  CellMemCache get _cellCache => _rowCache.cellCache;
 
   CellController({
-    required DatabaseCellContext cellContext,
-    required CellMemCache cellCache,
+    required this.viewId,
+    required FieldController fieldController,
+    required CellContext cellContext,
+    required RowCache rowCache,
     required CellDataLoader<T> cellDataLoader,
     required CellDataPersistence<D> cellDataPersistence,
-  })  : _cellContext = cellContext,
-        _cellCache = cellCache,
+  })  : _fieldController = fieldController,
+        _cellContext = cellContext,
+        _rowCache = rowCache,
         _cellDataLoader = cellDataLoader,
         _cellDataPersistence = cellDataPersistence,
-        _rowMetaListener = RowMetaListener(cellContext.rowId),
         _fieldListener = SingleFieldListener(fieldId: cellContext.fieldId),
-        _cacheKey = CellCacheKey(
-          rowId: cellContext.rowId,
-          fieldId: cellContext.fieldInfo.id,
-        ) {
-    _cellDataNotifier = CellDataNotifier(value: _cellCache.get(_cacheKey));
+        _cellDataNotifier =
+            CellDataNotifier(value: rowCache.cellCache.get(cellContext)) {
+    _startListening();
+  }
+
+  /// casting method for painless type coersion
+  CellController<A, B> as<A, B>() => this as CellController<A, B>;
+
+  /// Start listening to backend changes
+  void _startListening() {
     _cellListener = CellListener(
-      rowId: cellContext.rowId,
-      fieldId: cellContext.fieldInfo.id,
+      rowId: _cellContext.rowId,
+      fieldId: _cellContext.fieldId,
     );
 
-    /// 1.Listen on user edit event and load the new cell data if needed.
-    /// For example:
-    ///  user input: 12
-    ///  cell display: $12
+    // 1. Listen on user edit event and load the new cell data if needed.
+    // For example:
+    //  user input: 12
+    //  cell display: $12
     _cellListener?.start(
```

Comment:
make sure these listeners are disposed. Otherwise, it might cause memory leak

Reply:
Got it.

Reply:
I will take care of it in a follow up PR, since the changes will be quite big. I have made sure that the CellController in the blocs are always disposed.